src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -823,20 +827,153 @@ class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, gp: GridPartitioner) e
       while (indexedRows.hasNext && i < nRowsInBlock) {
         val indexedRow = indexedRows.next()
         assert(indexedRow.index == firstRowInBlock + i,
-          s"IndexedRow index ${indexedRow.index} in partition $pi does not equal RDD index ${firstRowInBlock + i}")
-        
+          s"IndexedRow index ${ indexedRow.index } in partition $pi does not equal RDD index ${ firstRowInBlock + i }")
+
         val data = indexedRow.vector.toArray // free on Spark DenseVector
         Memory.memcpy(bytes, 0, data, 0, data.length)
 
         var off = 0
         var blockCol = 0
         while (blockCol < gp.nBlockCols) {
           val n = gp.blockColNCols(blockCol) << 3
-          
+
           dosArray(blockCol).write(bytes, off, n)
-          
+
           off += blockSize << 3
-          blockCol+= 1
+          blockCol += 1
+        }
+        i += 1
+      }
+    }
+
+    dosArray.foreach(_.close())
+
+    Iterator.single(gp.nBlockCols) // number of blocks written
+  }
+}
+
+class WriteBlocksRDD2(vsm: VariantSampleMatrix, path: String, parentPartitionStarts: Array[Long], f: () => java.lang.Double, ec: EvalContext, gp: GridPartitioner) extends RDD[Int](vsm.sparkContext, Nil) {
```

Comment:
Hmm, taking a VSM might be a problem.  VSM isn't Serializable but RDDs need to be.  You might just take the things you need (which seem to include the sc as well as the variables below (rdd2, rowType, sampleIds, etc.))

Reply:
In fact, I'm not quite sure why this works.  Maybe because vsm is only a constructor and not a member because it isn't used in a method.  That rule always struck me as precarious.